[3.0] use ContainerAwareTrait

Tobion opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

Finished version of #12595 

This adds the first deprecation in symfony 3. We cannot deprecate ContainerAware in 2.8 as the alternative, the trait, cannot be used with php 5.3. So there would not be an upgrade path for people using an older php version

blanchonvincent committed:
Remove abstract class and use Interface+Trait

Conflicts:
	src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php
	src/Symfony/Bundle/SecurityBundle/Tests/Functional/Bundle/FormLoginBundle/Controller/LocalizedController.php
	src/Symfony/Bundle/SecurityBundle/Tests/Functional/Bundle/FormLoginBundle/Controller/LoginController.php
	src/Symfony/Component/DependencyInjection/ContainerAware.php

Tobion committed:
Remove deprecated ContainerAware class and make use of the trait in another class

Tobion commented:
@symfony/deciders this is ready.

aitboudad commented on src/Symfony/Component/DependencyInjection/ContainerAware.php:
```diff
@@ -11,10 +11,14 @@
+ * 
+ * @deprecated since version 3.0, to be removed in 4.0. Use ContainerAwareTrait with ContainerAwareInterface instead.
  */
 abstract class ContainerAware implements ContainerAwareInterface
```
we can use `ContainerAwareTrait` in this class too

Tobion replied:
good catch, changed

aitboudad commented:
:+1:

nicolas-grekas commented:
Shouldn't we deprecate in 2.8? IMHO yes, what do others think?

Tobion commented:
@nicolas-grekas please read the description.

WouterJ commented:
I think we should deprecate in 2.8:

 * Symfony 3 requires PHP 5.5
 * Symfony promises a very smooth (no changes required) upgrade process from 2.8 to 3.0 when having no deprecation notices in 2.8
 * This means applications have to run Symfony 2.8 on PHP 5.5 already
 * Which means they can use the trait (as it's available from PHP 5.4)

Actually, maybe we should even trigger a deprecation notice when PHP <5.5 in `Kernel#handle` or the like in Symfony 2.8.

nicolas-grekas commented:
Of course there would be an upgrade path: first upgrade to php5.5, which is already required. I agree with @wouterj arguments (but for notices, they should not be triggered conditionally upon the php version)

TomasVotruba commented:
:+1: Awesome

Tobion commented:
IMO developers should be able to upgrade their 2.x projects to 2.8 while avoiding any deprecation message. People stuck on PHP 5.3 would not be able to do so.
Anyway, I'm ok with whatever approach we take. Time for @fabpot to decide :)

Taluu commented on src/Symfony/Bridge/Doctrine/ManagerRegistry.php:
```diff
@@ -22,10 +22,7 @@
  */
 abstract class ManagerRegistry extends AbstractManagerRegistry implements ContainerAwareInterface
```
you should remove the implements here

jakzal commented:
status: reviewed

:+1:

fabpot commented:
IIRC, we agreed to not have any deprecation notices in 3.0. So, it's either 2.8 or 3.1.

Tobion commented:
Lets just merge #16424, ill rebase this one then.

nicolas-grekas commented:
While rebasing @Tobion you could also remove the uses of the VarDumperTestCase also (replaced by VarDumperTestTrait). Or I can do it, as you want.

Tobion committed:
[TwigBundle] use ContainerAwareTrait in commands

I made the class final because it is not meant to be extended. By using the trait, which uses protected visibility for the container property, it would otherwise make the container property part of of BC promise

nicolas-grekas commented:
Sibling PR on 2.8 to not break 3.0 compat: #16466

Tobion committed:
[VarDumper] replace VarDumperTestCase by trait

Tobion commented:
Rebased and I found two more classes (commands in the TwigBundle) where we can use the ContainerAwareTrait. I made those commands final because they is not meant to be extended. By using the trait, which uses `protected` visibility for the container property, it would otherwise make the container property part of of BC promise.
I also changed the VarDumper to use its trait.

nicolas-grekas commented on src/Symfony/Component/VarDumper/Tests/Test/VarDumpTestTraitTest.php:
```diff
@@ -9,8 +9,32 @@
+
+use Symfony\Component\VarDumper\Test\VarDumperTestTrait;
+
+class VarDumperTestTraitTest extends \PHPUnit_Framework_TestCase
```
a mistake that was here before but could be fixed here: the name of the file is wrong (VarDump**er**)

Tobion replied:
done

Tobion committed:
[VarDumper] fix filename

nicolas-grekas commented:
:+1:

fabpot commented:
Tests are broken.

nicolas-grekas commented:
The hard coded line info (63 to 63) needs to be updated (65 to 65)

Tobion commented:
Fixed tests

nicolas-grekas commented:
Thank you @Tobion.
